pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,290 @@
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+// schemerForGVK queries the openapi v3 discovery endpoint
+// and builds a custom openapi.GroupVersion to retrieve the schema
+// only for the GroupVersion of the given GroupVersionKind
+// It also stores the returned ETag associated with the GV for caching
+func (e *cachingUnstructuredExtractor) schemerForGVK(gvk schema.GroupVersionKind) (*CustomOAPIGroupVersion, error) {
+	data, err := e.dc.RESTClient().Get().
+		AbsPath("/openapi/v3").
+		Do(context.TODO()).
+		Raw()
+
+	if err != nil {
+		return nil, err
+	}
+
+	discoMap := &handler3.OpenAPIV3Discovery{}
+	err = json.Unmarshal(data, discoMap)
+	if err != nil {
+		return nil, err
+	}
+
+	gvPath := gvkToRelativeAPIPath(gvk)
+	oapiGV, ok := discoMap.Paths[gvPath]
+	if !ok {
+		return nil, fmt.Errorf("cannot find GroupVersion %q in discovery", gvPath)
+	}
+
+	parse, err := url.Parse(oapiGV.ServerRelativeURL)
+	if err != nil {
+		return nil, err
+	}
+	useClientPrefix := strings.HasPrefix(oapiGV.ServerRelativeURL, "/openapi/v3")
+	etag := parse.Query().Get("hash")
+
+	return NewCustomOAPIGroupVersion(e.dc.RESTClient(), oapiGV, useClientPrefix, etag), nil
+}
+
+// gvkToRelativeAPIPath constructs the OpenAPI path for the given GVK
+func gvkToRelativeAPIPath(gvk schema.GroupVersionKind) string {
+	if gvk.Group == "" {
+		return "api/" + gvk.Version
+	}
+	return "apis/" + gvk.GroupVersion().String()
+}
+
+// parserFnForSchemer returns a BuildParserFn that builds a new *GvkParser
+// from the OpenAPI GroupVersion schema client
+func parserFnForSchemer(schemer *CustomOAPIGroupVersion) func() (*GvkParser, error) {
+	return func() (*GvkParser, error) {
+		s, err := schemer.Schema("application/json")
+		if err != nil {
+			return nil, err
+		}
+		var oapi spec3.OpenAPI
+		if err := json.Unmarshal(s, &oapi); err != nil {
+			return nil, err
+		}
+
+		specs := map[string]*spec.Schema{}
+		for k, v := range oapi.Components.Schemas {
+			specs[k] = v
+		}
+		return NewGVKParser(specs, false)
+	}
+}
+
+func (e *cachingUnstructuredExtractor) objectForGVK(gvk schema.GroupVersionKind) (*typed.ParseableType, error) {
+	schemer, err := e.schemerForGVK(gvk)
+	if err != nil {
+		return nil, err
+	}
+
+	parser, err := e.cache.ParserForGV(gvk.GroupVersion().String(), schemer.etag, parserFnForSchemer(schemer))
```

Comment:
if we didn't use `GVKParserCacheManager` but e.g. https://pkg.go.dev/github.com/hashicorp/golang-lru/v2#section-readme, couldn't this be simplified into:

```go
parserTuple, ok := e.cache.Get(gvk.GroupVersion())
if !ok || parserTuple.etag != schemer.etag {
    parserTuple.parser = parserFnForSchemer(schemer)
    parserTuple.etag = schemer.etag
    e.cache.Add(gvk.GroupVersion(), parserTuple)
}
return parserTuple.parser
```

we should also probably have a special case for etag="" if that ever would happen to fetch every time and not cache such a thing.

Reply:
> if we didn't use GVKParserCacheManager
Just to be on the same page, I assume that you refer to `GVKParserCache` here as the suggested code has the GV as the key.

there are two layers of storage involved 
`GvkParserCacheManager` is a mere catalog of caches per providerconfig. You get the relevant cache upon `Connect()` for your providerconfig. The keys, and values (pointer-wise) does not change once added, so no access to stale data happens, nor invalidation (not counting the case providerconfig getting deleted here)

`GVKParserCache` is a collection of parsers per GV, dedicated to a single providerconfig.

Back to the question:
I think there is a theoretical possibility of this handling it in a different way.

Since they do not share the same critical section, between the `cache.Get` and `cache.Add` , the entry can be filled by another routine.
Current change implements 
- `RLock` first, return if exists
- if not found, `Lock`
- check again for existence
- if cache was populated in between, return it
- if not, calculate new entry in critical section and return it
 
Not sure how this benefits in practical scenarios but maybe we can do e.cache.PeekOrAdd() instead of Add().
The difference would be that, the parser was already built, if the PeekOrAdd results in a successful peek.

The disadvantage in the current state of the change is that we build the parser in the critical section.

Reply:
I'd opt for the simplest and most understandable solution, over a hyper-optimized and hard to understand one.
Unix principle: Clarity over cleverness.
These schemas shouldn't take that long to process, so probably just one normal mutex is fine, and don't think it'd be that bad to populate in the critical section.